src/validations/html/rules/MarkupOutsideContainer.js

```diff
@@ -0,0 +1,75 @@
+// This rule checks for any unexpected tags or text that
+// don't get caught by other validators, specifically:
+//
+// * Any tags or text outside the <html> element
+// * Any text that is inside the <html> element but not part
+//   of the <head> or <body> tags
+
+import findLastIndex from 'lodash-es/findLastIndex';
+
+import Code from './Code';
+
+export default class MismatchedTag {
+  constructor() {
+    this._openTagStack = [];
+    this._locationsWithOutsideTags = [];
+  }
+
+  openTag(location, {name}) {
+    if (
+      this._isOutsideAcceptableContainer() &&
+      name !== 'html' &&
+      name !== 'head' &&
+      name !== 'body'
+    ) {
+      this._locationsWithOutsideTags.push({
+        location,
+        type: `<${name}> tags`,
+        outsideTag: this._openTagStack.length === 0 ? 'html' : 'body',
+      });
+    }
+
+    this._openTagStack.push({location, name});
+  }
+
+  closeTag(location, name) {
+    const openIndex = findLastIndex(
+      this._openTagStack,
+      openTag => openTag.name === name,
+    );
+
+    const isOpened = openIndex >= 0;
+    if (isOpened) {
+      this._openTagStack.splice(openIndex);
+    }
+  }
+
+  text(location, text) {
+    const trimmedText = text.trim();
+    if (trimmedText && this._isOutsideAcceptableContainer()) {
+      this._locationsWithOutsideTags.push({
+        location,
+        type: 'text',
```

Comment:
I don’t have a strong opinion as to exactly what the right place is to make the split, but we should avoid having the word `text` be bold/monospace in the error message (the actual copy that it currently produces happens to be fine)

Reply:
@outoftime I've been thinking about the actual error messages a bit more, and I think my preference is to show the following two messages:

```
You are not allowed to write text outside of the <body> tag
You are not allowed to place `<p>` tags outside of the <body> tag
```

I think it will always indicate that you should be placing it inside the `<body>` tag. I think this also simplifies the error message names to `invalid-text-outside-body` and `invalid-tag-outside-body`

Thoughts on this naming?